src/Chemistry.hpp

```diff
@@ -0,0 +1,121 @@
+#ifndef CHEMISTRY_HPP_ // NOLINT
+#define CHEMISTRY_HPP_
+//==============================================================================
+// TwoMomentRad - a radiation transport library for patch-based AMR codes
+// Copyright 2020 Benjamin Wibking.
+// Released under the MIT license. See LICENSE file included in the GitHub repo.
+//==============================================================================
+/// \file CloudyCooling.hpp
+/// \brief Defines methods for interpolating cooling rates from Cloudy tables.
+///
+
+#include <limits>
+
+#include "AMReX.H"
+#include "AMReX_BLassert.H"
+#include "AMReX_Extension.H"
+#include "AMReX_GpuQualifiers.H"
+
+#include "hydro_system.hpp"
+#include "radiation_system.hpp"
+
+#include "burn_type.H"
+#include "burner.H"
+#include "eos.H"
+#include "extern_parameters.H"
+
+namespace quokka::chemistry
+{
+template <typename problem_t> void computeChemistry(amrex::MultiFab &mf, const Real dt_in, const Real max_density_allowed)
+{
+	const Real dt = dt_in;
+
+	for (amrex::MFIter iter(mf); iter.isValid(); ++iter) {
+		const amrex::Box &indexRange = iter.validbox();
+		auto const &state = mf.array(iter);
+
+		amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			const Real rho = state(i, j, k, HydroSystem<problem_t>::density_index);
+			const Real Eint = state(i, j, k, HydroSystem<problem_t>::internalEnergy_index);
```

Comment:
This was problematic in FLASH, because, often when shocks were present, subtraction of internal energy from the total energy resulted in very small or even negative internal energies, depending on the precision with which they were stored and subtraction was done. This would then make the chemistry barf. 

Since quokka stores the momentum (or is it specific momentum?)?

Reply:
The issue with negative internal energies is *not* due to floating-point precision, but due to discretization error. At high Mach numbers, the spatial resolution requirements formally increase as $\mathcal{M}$ (see https://arxiv.org/abs/0704.2608).

Reply:
I just realised something: in Microphysics, the internal energy I calculate is the specific internal energy. I find it the way FLASH finds it (using equation 16.6 of https://flash.rochester.edu/site/flashcode/user_support/flash4_ug_4p7/node108.html). FLASH also used specific internal energy, and the kinetic energy in FLASH was `ekin = 0.5*(vx^2 + vy^2 + vz^2)`. So, energies in FLASH are energy per unit mass (km^2 s^-2).

I just checked `hydro_system.hpp` in Quokka, and Quokka uses  `ekin = 0.5*rho*(vx^2 + vy^2 + vz^2)`. So, Quokka energies are (g/cm^3)*(km^2 s^-2). This means there will be an inconsistency between internal energy calculated by chemistry (via microphysics) and hydro (in quokka)?

Reply:
You have to convert between specific internal energy (SIE) and internal energy by dividing/multiplying by the gas density.

Reply:
In the code, I have distinguished between SIE and internal energy by using lowercase `e` for SIE and uppercase `Eint` for internal energy (where $E_{int} = \rho e$).

Reply:
Yeah. I was wondering if I should calculate Eint in microphysics in the same units as quokka. But all microphysics EOS use specific internal energy and we want to keep all Microphysics EOS consistent. So using a conversion factor between microphysics EOS and quokka EOS seems like the best choice.

Reply:
Yes, I agree.